backends/p4tools/modules/smith/common/declarations.cpp

```diff
@@ -746,4 +752,54 @@ IR::ParameterList *DeclarationGenerator::genParameterList() {
     return new IR::ParameterList(params);
 }
 
+/// @brief Check if the target is bmv2.
+bool DeclarationGenerator::isBmv2Target() {
+    if (target().get().spec.deviceName.compare("bmv2") == 0) {
+        return true;
+    }
+    return false;
+}
+
+/// @brief Generate a parameter of some primitive type.
+// TODO(zzmic): Figure out whether "primitive type" is the correct term to use.
+IR::Parameter *DeclarationGenerator::genPrimitiveTypeParameter(bool if_none_dir) {
```

Comment:
I believe you could achieve the same effect by setting the percentages appropriately. Then you do not need to use a separate method. You could make the percentages a parameter of the function.

Reply:
Yes. However, a concern I have is when I used `TyperefProbs typePercent` while taking PCT.PARAMETER_NONEDIR_BASETYPE_BIT, PCT.PARAMETER_NONEDIR_BASETYPE_SIGNED_BIT, and PCT.PARAMETER_NONEDIR_BASETYPE_INT (or PCT.PARAMETER_BASETYPE_BIT, PCT.PARAMETER_BASETYPE_SIGNED_BIT, and PCT.PARAMETER_BASETYPE_INT), it triggered a bunch of "not initialized" warning (for other members) while building. I'm not sure if that is permissible. Thus, I defined another struct called `PrimitiveTyperefProbs` that only requires PCT.PARAMETER_NONEDIR_BASETYPE_BIT, PCT.PARAMETER_NONEDIR_BASETYPE_SIGNED_BIT, and PCT.PARAMETER_NONEDIR_BASETYPE_INT to be initialized.

I have two alternatives in my mind to address the issue you pointed out. The first alternative is to employ overloading, defining two versions of the function and taking different types of the parameter (either `TyperefProbs` or `PrimitiveTyperefProbs`). But this, in turn, requires two functions to be defined (i.e., a separate function needs to be created). Another option is to use templates and template specialization. Is this more recommended?

Reply:
It doesn't need to be that complicated, how were you getting the warnings or errors? Ideally, you would do something like 
```
    TyperefProbs typePercent = {
        PCT.FUNCTIONDECLARATION_BASETYPE_BIT,    PCT.FUNCTIONDECLARATION_BASETYPE_SIGNED_BIT,
        PCT.FUNCTIONDECLARATION_BASETYPE_VARBIT, PCT.FUNCTIONDECLARATION_BASETYPE_INT,
        PCT.FUNCTIONDECLARATION_BASETYPE_ERROR,  PCT.FUNCTIONDECLARATION_BASETYPE_BOOL,
        0, PCT.FUNCTIONDECLARATION_DERIVED_ENUM,
        0,  0,
        0,  P0,
        0,   0,
        0,
    };
```